core/src/main/scala/filodb.core/downsample/ChunkDownsampler.scala

```diff
@@ -0,0 +1,272 @@
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+
+    val vecPtr = chunkset.vectorPtr(colIds(0))
+    val colReader = part.chunkReader(colIds(0), vecPtr)
+    colReader.asDoubleReader.count(vecPtr, startRow, endRow)
+  }
+}
+
+/**
+  * Downsamples by calculating min of values in one column
+  */
+case class MinDownsampler(override val colIds: Array[Int]) extends DoubleChunkDownsampler {
+  require(colIds.length == 1, s"Min downsample requires only one column. Got ${colIds.length}")
+  override val name: String = "dMin"
+  override def downsampleChunk(part: TimeSeriesPartition,
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+
+    val vecPtr = chunkset.vectorPtr(colIds(0))
+    val colReader = part.chunkReader(colIds(0), vecPtr)
+
+    var min = Double.MaxValue
+    var rowNum = startRow
+    val it = colReader.iterate(vecPtr, startRow).asDoubleIt
+    while (rowNum <= endRow) {
+      val nextVal = it.next
+      if (!JLDouble.isNaN(nextVal)) min = Math.min(min, nextVal)
+      rowNum += 1
+    }
+    min
+  }
+}
+
+/**
+  * Downsamples by calculating max of values in one column
+  */
+case class MaxDownsampler(override val colIds: Array[Int]) extends DoubleChunkDownsampler {
+  require(colIds.length == 1, s"Max downsample requires only one column. Got ${colIds.length}")
+  override val name: String = "dMax"
+  override def downsampleChunk(part: TimeSeriesPartition,
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+
+    val vecPtr = chunkset.vectorPtr(colIds(0))
+    val colReader = part.chunkReader(colIds(0), vecPtr)
+
+    var max = Double.MinValue
+    var rowNum = startRow
+    val it = colReader.iterate(vecPtr, startRow).asDoubleIt
+    while (rowNum <= endRow) {
+      val nextVal = it.next
+      if (!JLDouble.isNaN(nextVal)) max = Math.max(max, nextVal)
+      rowNum += 1
+    }
+    max
+  }
+}
+
+/**
+  * Downsamples by calculating average from average and count columns
+  */
+case class AvgAcDownsampler(override val colIds: Array[Int]) extends DoubleChunkDownsampler {
+  require(colIds.length == 2, s"AvgAc downsample requires column ids of avg and count. Got ${colIds.length}")
+  override val name: String = "dAvgAc"
+  val avgCol = colIds(0)
+  val countCol = colIds(1)
+  override def downsampleChunk(part: TimeSeriesPartition,
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+    val avgVecPtr = chunkset.vectorPtr(avgCol)
+    val avgColReader = part.chunkReader(avgCol, avgVecPtr)
+    val cntVecPtr = chunkset.vectorPtr(countCol)
+    val cntColReader = part.chunkReader(countCol, cntVecPtr)
```

Comment:
You might want to check that the count column is in fact a DoubleColumn.  It might not be, or the user might make a mistake in configuration.

Reply:
This is where I think simplifying with just double column (for now) makes configuration as well as implementation simple. 

Here is where things gets murky:

* Can count downsample be a long or double?
* Should sum of count columns be a long or double?
* Should sum then have a long version too?
* Can sum and count types interchange ?

It is just simple to deal with one input type (at least for now) and cross the bridge of dealing with multiple types later. Let's discuss in person tomorrow. This thread may get long.